[stm32] Add FDCAN support for H7 series

WasabiFan opened the pull request:
This PR updates the stm32-fdcan driver to support the STM32H7 series.

Note that I don't have any prior knowledge of STM32's CAN IP so I am learning it as I go. I'd appreciate a fact-check on what I've done here.

High-level differences between the H7 FDCAN IP and prior families:
- The message RAM size has been increased from 1.6K to 10K
- The fixed message RAM layout has been replaced with a set of base address and count registers. SW configures the location and size of each message RAM region.
- SW can configure the number of data bytes in each frame entry in the TX and RX structures. It now resets to 8 instead of 64.
- There are some new features I haven't investigated (TTCAN, dedicated RX buffers, maybe others)
- Misc registers have been re-arranged pursuant to the above.
- The interrupt line select bits have improved granularity.

This PR does the following:
- Update existing config register writes for the fields' new locations
- Add initialization logic to configure the message RAM layout and message size
- Add a new configuration struct and associated lbuild options to select the size of the message RAM structures
  - This replaces the hard-coded message ram layout computation.
  - For H7, these values are configurable and I have given them slightly higher defaults to take advantage of the larger message RAM. For other families, I have left these set to the fixed values to match the old behavior.
- Updates the fdcan test so it can fill the new buffer space
- Adds an example to get CI compilation coverage

Validation:
- On an STM32H753 I have manually confirmed basic non-FD CAN TX and RX on both FDCAN instances.
  - The fdcan test passes with a caveat I'll note in a comment.
  - Note that the board I'm developing on is not one currently supported by modm (I may open a separate PR for that). The Nucleo example I have included here is untested.
- I'm marking this PR as a draft for now as I need to do some more testing of filters and CAN-FD.
- I've compiled the examples for all boards.

Pending TODOs:

- [x] Fix CAN test error on STM32H753 (last message dropped)
- [x] Cherry-pick https://github.com/chris-durand/modm/commit/ddf3d179e0a8c89047f86770c009fb4f04a52ecf
- [x] Disable HW prioritized queue mode
- [x] Validate CAN-FD features on STM32
- [x] Validate all functionality on SAM to confirm no regressions
- [x] Validate all functionality on a non-H7 STM32 (e.g. STM32G4) to confirm no regressions

I would appreciate if someone has a Nucleo STM32H7 board and/or a non-H7 board to double check my testing and confirm I haven't broken the other chip families. I have only tested on H7.

chris-durand committed:
[stm32] Allow using hardware buffers only in FDCAN driver

WasabiFan committed:
[stm32] Add FDCAN support for H7 series

The H7 series supports a user-configured message RAM layout, rather than
the static layout of past STM32 CAN IPs. The new message RAM is
identical in structure and configurability to the ATSAM FDCAN IP. This
commit merges the STM32 FDCAN and ATSAM FDCAN message RAM
implementations.

For H7 it also introduces lbuild parameters so the user can choose the amount
of space available for each RX queue, each TX queue, and filters. This is
independent for each FDCAN IP instance, although they share the same 10k RAM.

WasabiFan commented:
The H7 reference manual (and even the older G4 manual) have what I believe to be oversights/errors which make me a bit unsure about the data size configuration and message RAM layout. For example, the H7 manual says this about TX queue entries:

> A Tx queue buffer allocates four 32-bit words in the message RAM. Therefore the start
address of the next available (free) Tx queue buffer is calculated by adding four times the Tx
queue put index FDCAN_TXFQS.TFQPI (0 … 31) to the Tx buffer start address
FDCAN_TXBC.TBSA.

I believe this to be wrong. If TXESC.TBDS is set such that CAN messages are 8 bytes, it is true. But if it is set to any other value, there needs to be space for those extra bytes. Emperically it does seem that setting TBDS to a higher value grows the queue entries accordingly.

For TX event entries, which are structured similarly, it does say what I believe to be correct:

> The Tx event FIFO can be configured to a maximum of 32 elements. The Tx event FIFO
element is described in Tx FIFO. Depending on the configuration of the element size
(FDCAN_TXESC), between two and sixteen 32-bit words (Tn = 3 ..17) are used for storage
of a CAN message data field.

WasabiFan commented on examples/nucleo_h723zg/can/main.cpp:
```diff
@@ -0,0 +1,82 @@
+/*
```
This example has been compiled but not run. I would appreciate a test run on one of the standard nucleo/discovery boards by someone who owns one.

WasabiFan commented on src/modm/platform/can/stm32-fdcan/can.cpp.in:
```diff
@@ -192,13 +195,67 @@ modm::platform::Fdcan{{ id }}::initializeWithPrescaler(
 
 	// Tx buffer: queue mode
-	{{ reg }}->TXBC = FDCAN_TXBC_TFQM;
+	{{ reg }}->TXBC |= FDCAN_TXBC_TFQM;
```
Can anyone comment on why we enable queue mode? It leads to a weird behavior where the in-memory/SW-managed queue is a true FIFO but the HW is arbitrating according to message ID priority. In fact, this breaks the fdcan buffer test, which expects to receive messages in the order they were enqueued.

I see some related discussion here: https://github.com/modm-io/modm/discussions/710.

My opinion is that this behavior is undesireable so long as the SW-managed queue remains. I would _prefer_ to disable it.

chris-durand replied:
The behaviour is rather weird with the software queue and disabling it seems reasonable. 

Furthermore, we should allow to completely bypass the software queues. In that case I'd prefer to have the arbitration. With the 10 kB message RAM on H7 using the hardware queues only is sufficient for many applications. The code is already there in the driver, we only need to allow users to configure the software buffers as zero length: https://github.com/chris-durand/modm/commit/ddf3d179e0a8c89047f86770c009fb4f04a52ecf

chris-durand replied:
Should we somehow add an option to enable queue mode? In case the additional software queue isn't used it seems to be a useful feature.

WasabiFan commented on src/modm/platform/can/stm32-fdcan/message_ram.hpp:
```diff
@@ -28,9 +29,108 @@
+	constexpr uint32_t rxFifo0SectionOffset() const { return extendedFilterListSectionOffset() + extendedFilterListSectionWords(); }
+	constexpr uint32_t rxFifo1SectionOffset() const { return rxFifo0SectionOffset() + rxFifo0SectionWords(); }
+	constexpr uint32_t txEventSectionOffset() const { return rxFifo1SectionOffset() + rxFifo1SectionWords(); }
+	constexpr uint32_t txFifoSectionOffset() const { return txEventSectionOffset() + txEventSectionWords(); }
```
TODO: verify that these functions are being evaluated at compile time

WasabiFan commented on src/modm/platform/can/stm32-fdcan/module.lb:
```diff
@@ -62,6 +107,8 @@ class Instance(Module):
         properties["shared_irq_it0"] = False
         properties["shared_irq_it1"] = False
 
+        properties["message_ram_length_words"] = 2560 if device.identifier.family == "h7" else 1696
```
I didn't see a sensible place in the Cube database to pull this number from, but I may not have been looking in the right place.

chris-durand replied:
Is the size of 1696 words really correct here? According to the G4 reference manual it's 212 words (0.8 kB) per FDCAN instance.

WasabiFan replied:
Good catch. I have no idea where I got this number.

chris-durand commented:
@salkinium I'll take a look at this in detail soon. I'm currently traveling within Asia and didn't have access to a computer for a week. 

I also need the H7 CAN support at work and will have to ship it in a product. In a first attempt to implement this I've reused the message RAM definitions [from the ATSAM MCAN driver](https://github.com/modm-io/modm/blob/develop/src/modm/platform/can/sam-mcan/message_ram.hpp) which is same hardware IP. You can find the draft code [here](https://github.com/chris-durand/modm/tree/feature/stm32h7_can). It's tested in hardware on a custom H723ZG board. 

The ATSAM driver message RAM class also takes a config non-type template parameter, similar to what was done in this PR. I'm wondering if it could be worth unifying both that we don't have maintain two implementations of the same thing. @WasabiFan Could you check whether that makes sense and looks feasible?

WasabiFan commented:
Thanks for taking a look.

I hadn't realized the SAM CAN IP was identical. I had assumed there was just some shared lineage. I'll take a closer look after work.

At first glance the message RAM looks to be the same and we ought to be able to share it. In fact, the config registers seem identical too, so we could consider sharing the whole driver. It would require some templating for the register defines and the interrupts would still be special-cased.

The main difference I see so far is message RAM addressing. The STM32H7 as I understand it uses a dedicated SRAM for the message RAM which is shared between the N FDCAN IP instances. Addresses loaded into the region start address fields are offsets into the 10k SRAM. Meanwhile, the SAM datasheet says "Multiple MCANs May Share the Same Message RAM" which seems to mean it addresses shared system memory directly. The modm driver for SAM MCAN allocates a local buffer in the CAN driver class and points the CAN IP to that. The MCAN driver does not need to apply an explicit allocation policy for the shared RAM. My PR for H7 currently allows the user to configure their message RAM region sizes and then computes offsets for each FDCAN instance so they sit contiguously in the SRAM. The MCAN message RAM class would need a small amount of rework to support this.

@chris-durand is the request to share the MessageRam class or the whole MCAN/FDCAN driver?

WasabiFan commented:
I've updated the PR with an implementation which shares the MessageRam class between SAM and STM32 FDCAN. There's a test failure I'll look into later; TODOs now posted in the PR description. I'd appreciate a review of this to get some feedback on the overall change in the meantime.

Note that the SAM CAN driver takes configuration from the user via a template parameter, while the STM32 implementation takes it from lbuild parameters. This is because the STM32 FDCAN instances share a hardware RAM. Each must know how much space the prior instance(s) take up.

I'm not fond of the size parameters in the message RAM config struct storing total element size (including header) rather than just size of the message body. The user shouldn't see or care how large the header is. However, the existing SAM driver does this and I am worried about introducing a subtle foot-gun if I redefine it now for users already configuring the SAM FDCAN driver by hand. So I've left it as-is.

WasabiFan committed:
[stm32] Default FDCAN to FIFO mode instead of Queue Mode

Queue Mode will send CAN frames in priority order according to their
arbitration ID. This is inconsistent with modm's software-managed queue
which is a strict FIFO. To make the behavior of the driver consistent
with an end-to-end FIFO order, we disable Queue Mode by default.

There is a new lbuild option, "tx_hw_queue_mode", which allows the user
to opt back into the old behavior if desired. This is mostly intended
for if they also set "buffer.tx" to zero and optionally increase
"message_ram.tx_fifo_elements".

Note that the fdcan unit test was implicitly assuming FIFO order. This
change also fixes that test for larger HW TX buffer sizes.

chris-durand commented on src/modm/board/nucleo_h723zg/board.hpp:
```diff
@@ -74,8 +74,8 @@ struct SystemClock
-	static constexpr uint32_t Can1 = Apb1;
-	static constexpr uint32_t Can2 = Apb1;
+	static constexpr uint32_t Fdcan1 = Apb1;
+	static constexpr uint32_t Fdcan2 = Apb1;
```
`Apb1` doesn't look right. Possible FDCAN clock sources are Hse, Pll1Q and Pll2Q. Default after reset is Hse, which is an 8 MHz clock on the Nucleos.

chris-durand commented on src/modm/board/nucleo_h743zi/board.hpp:
```diff
@@ -70,8 +70,8 @@ struct SystemClock
 
-	static constexpr uint32_t Can1 = Apb1;
-	static constexpr uint32_t Can2 = Apb1;
+	static constexpr uint32_t Fdcan1 = Apb1;
+	static constexpr uint32_t Fdcan2 = Apb1;
```
same as above

chris-durand commented on src/modm/platform/can/stm32-fdcan/can.hpp.in:
```diff
@@ -33,18 +34,43 @@ namespace modm::platform
+ * but arbitrated according to ID priority within the HW queue.
  *
  * ## Filter
+%% if target["family"] == "h7"
  * Up to 28 filters can be defined for 11-bit IDs, up to 8 filters for 29-bit IDs.
+%% else
```
The `if` is the wrong way around. The H7 and non-H7 sections are swapped.

chris-durand commented on src/modm/platform/clock/stm32/module.lb:
```diff
@@ -105,6 +105,16 @@ def build(env):
         properties["cfgr2"] = "CFGR"
 
+    if target.family in ["h7"]:
+        if target.name in ["a0", "a3", "b0", "b3"]:
```
```suggestion
        if target.name in ["a3", "b0", "b3"]:
```

There are no H7A0 devices

WasabiFan replied:
Manually applied this suggestion and fixed the other instances of the same. I was wondering what was going on here 🙂

chris-durand reviewed (commented):
Thank you very much, very nice! I'll run it on hardware on Wednesday when I'm back home.

chris-durand commented:
> I'm not fond of the size parameters in the message RAM config struct storing total element size (including header) rather than just size of the message body. The user shouldn't see or care how large the header is. However, the existing SAM driver does this and I am worried about introducing a subtle foot-gun if I redefine it now for users already configuring the SAM FDCAN driver by hand. So I've left it as-is.

I agree but I wouldn't want to fix it in a way that silently breaks user code. It could theoretically be done by introducing a new struct and deprecating the old, but I don't think it's worth the effort.

WasabiFan committed:
[stm32] Remove references to nonexistent 'a0' STM32H7 chip

WasabiFan commented:
I believe I've updated the PR with the review feedback. I also autosquashed all my fixup commits since I re-arranged some hunks between commits.

The ordering of the HW TX queue is now configurable.

To ensure I haven't regressed existing behavior, I'd appreciate a test in hardware on a) a non-H7 STM32 and b) an ATSAM board. c) one of the Nucleos whose board.hpp has been retrofitted (h723zg or h743zi) would be nice too if possible.

I still have some more validation which I will do on my H753 third-party board.

chris-durand commented:
> To ensure I haven't regressed existing behavior, I'd appreciate a test in hardware on a) a non-H7 STM32 and b) an ATSAM board. c) one of the Nucleos whose board.hpp has been retrofitted (h723zg or h743zi) would be nice too if possible.

The unit test fails on both the G474RE and the H723ZG Nucleo board. I'll investigate tomorrow.

WasabiFan commented:
Sorry, I still haven't fixed that. I see the same on my H7 and haven't looked into it. The final message is being dropped. Those tests were passing before I swapped the message RAM implementation so I assume it's a legitimate bug.

WasabiFan commented:
The test is fixed. All FDCAN tests are now passing on my STM32H753. It was a dumb off-by-one bug I introduced in the fdcan buffer test when I made it vary the number of messages based on available message RAM/queue space.

WasabiFan commented:
I did a sanity check of FD features (various larger frames up to 64b, higher data bit rate) with a physical loopback on the H753 and don't see any issues.

chris-durand commented:
> The test is fixed. All FDCAN tests are now passing on my STM32H753. It was a dumb off-by-one bug I introduced in the fdcan buffer test when I made it vary the number of messages based on available message RAM/queue space.

Nice!

There are still issues on the Nucleo boards. On the H7 the peripheral doesn't work at all clocked from 8 MHz HSE. I assume the frequency is just too low. When I switch the clock mux to Pll1Q the test passes. If we do that we'll have to increase the Pll1Q clock divider from 4 to 5 to not exceed the FDCAN maximum clock of 125 MHz. I'll check if it works on my custom board with a 24 MHz HSE crystal.

The G4 test fails on line 63, but it even does so without the changes from this PR. A message is received although it should be rejected. If I consume the message before the check all remaining tests pass. @WasabiFan Do you have any idea what could be wrong here?

> FAIL: fdcan:63 : true == false
FAIL: fdcan:104 : 1801 == 18193
FAIL: fdcan:106 : 0 == 1
FAIL: fdcan:108 : 0 == 1
FAIL: fdcan:106 : 1 == 2
FAIL: fdcan:108 : 1 == 2
FAIL: fdcan:108 : 52 == 2
FAIL: fdcan:106 : 2 == 3
FAIL: fdcan:108 : 2 == 3
FAIL: fdcan:108 : 2 == 3
FAIL: fdcan:108 : 86 == 3
FAIL: fdcan:106 : 3 == 4
FAIL: fdcan:108 : 3 == 4
FAIL: fdcan:108 : 3 == 4
FAIL: fdcan:108 : 3 == 4
FAIL: fdcan:108 : 239 == 4
FAIL: fdcan:106 : 4 == 5
FAIL: fdcan:108 : 4 == 5
FAIL: fdcan:108 : 4 == 5
FAIL: fdcan:108 : 4 == 5
FAIL: fdcan:108 : 4 == 5
FAIL: fdcan:108 : 0 == 5

chris-durand commented:
On my custom board the 24 MHz HSE clock appears to work fine.

WasabiFan commented:
Re: the clocking, my H7 board has a 16MHz crystal and I confirmed FDCAN works both off that and off a PLL. Although, I'm only now realizing that I was overclocking it by a factor of 2x when running from the PLL because I misread the Fmax table. It's impressive that it still worked.

For those Nucleo boards, can we switch it to use PLL2Q instead?

Re: the G4 test, can you try commenting out the `setStandardFilter` so there are no standard acceptance criteria? I guess also try disabling the extended filter to confirm that isn't somehow affecting the standard ones. Accepting a message which shouldn't be accepted is a weird failure mode since it implies the acceptance filter is present and recognized but overly broad.

I took another look for differences between the G4 and H7 filter handling. I don't see any functional differences without explicit opt-in. I do notice is that the mask (ID2) layout is documented in a way that is incorrect or at least heavily confusing in the H7 version, but I interpret it to be poor documentation rather than a design difference. The SAM datasheet matches the G4's interpretation of these fields.

![Screenshot 2024-09-21 011312](https://github.com/user-attachments/assets/08af2ab6-6c81-4b6f-8390-86972afd363e)

WasabiFan commented:
Reflecting on the message RAM size review comment posted above: there's a legitimate functional bug I need to fix. The way I implemented computation of each peripheral's message RAM base is wrong for the fixed layout (non-H7) when not all instance modules are included in lbuild. I'll need to make the address computation independent of the enabled peripherals.

WasabiFan commented:
Yesterday I fixed the message RAM allocation bug. That doesn't explain the G4 failure. It would be good to run the tests on a G4 with FDCAN0 disabled to confirm FDCAN1 still works. If you'd like me to take a closer look at the failure I can order a G4 Nucleo.

chris-durand commented:
> If you'd like me to take a closer look at the failure I can order a G4 Nucleo.

I've just found the reason for the G4 failure. After power-on the whole FDCAN SRAM is filled with garbage values and we never zero it:

![Screenshot_20240923_210154](https://github.com/user-attachments/assets/7ef885ab-2f0a-4716-a293-eb70ffcc1b37)

On my G4 Nucleo Board one memory location happens to reliably produce a valid standard CAN filter entry accepting almost all messages. When I flash a program to memset the whole message RAM to 0s and then reflash the original unit test binary all tests pass until I power-cycle the board.

This bug affects all devices, not only G4. I'll push a fix tomorrow.
Someone might rely on the fact that `Fdcan::initialize()` doesn't reset the message RAM content, e.g. if the intention is to restore the correct timings after a system clock change. Thus, I'd only zero it on the first call and store a bool flag inside the Fdcan class.

> It would be good to run the tests on a G4 with FDCAN0 disabled to confirm FDCAN1 still works

The unit test still succeeds when I run it on the second or third instance. At first I was totally confused when I looked at the reference manual, but it turns out the memory map in the documentation is entirely wrong. It suggests the message RAMs for each instance are 1024 bytes large and there are unused areas at the end of each section:

![Screenshot_20240923_222858](https://github.com/user-attachments/assets/a463be35-02dd-43ec-8c38-291c67a1cb89)

The ST HAL implementation agrees with our understanding that the message RAM consists of three contiguous 848 byte blocks.

WasabiFan commented:
> I've just found the reason for the G4 failure. After power-on the whole FDCAN SRAM is filled with garbage values and we never zero it:

Nice find! The ST driver implementation does zero out the message RAM, I guess that's why. I had implemented this locally in modm as part of debugging a different issue but didn't think to keep it...

>  Someone might rely on the fact that Fdcan::initialize() doesn't reset the message RAM content, e.g. if the intention is to restore the correct timings after a system clock change. Thus, I'd only zero it on the first call and store a bool flag inside the Fdcan class.

Seems reasonable to me.

> The ST HAL implementation agrees with our understanding that the message RAM consists of three contiguous 848 byte blocks.

The message RAM section also agrees with this interpretation. The memory map seems like complete nonsense to me ☹️ 

![image](https://github.com/user-attachments/assets/ffe9aba8-ebc0-4766-b2d2-71fe93d80561)

chris-durand committed:
[stm32] Fix initialization of FDCAN message RAM

After power-on the FDCAN message RAM is filled with non-zero invalid
data. In some cases the CAN filter region contains well-formed entries
which cause inadvertent reception or rejection of messages.

To prevent this message RAM sections are now zeroed out on the first
initialization of the respective FDCAN instance.

chris-durand commented on src/modm/platform/can/sam-mcan/can.hpp:
```diff
@@ -83,16 +84,9 @@ class McanDriver : public ::modm::Can
-	static_assert(MessageRam::Size <= 4352*4, "Max message ram size is 4352 words.");
-
-	static inline std::array<uint32_t, MessageRam::Size/4> modm_aligned(4)
+	static_assert(mrc.totalSectionWords() <= 4352*4, "Max message ram size is 4352 words.");
```
`*4` seems to be unneccessary. `totalSectionWords()` should be in words, not bytes.

chris-durand commented on src/modm/platform/can/sam-mcan/can.hpp:
```diff
@@ -83,16 +84,9 @@ class McanDriver : public ::modm::Can
-	static inline std::array<uint32_t, MessageRam::Size/4> modm_aligned(4)
+	static_assert(mrc.totalSectionWords() <= 4352*4, "Max message ram size is 4352 words.");
+
+	static inline std::array<uint32_t, mrc.totalSectionWords()/4> modm_aligned(4)
```
```suggestion
	static inline std::array<uint32_t, mrc.totalSectionWords()> modm_aligned(4)
```
same here with `/4`

chris-durand commented:
@WasabiFan I have fixed the initialization bug and changed the FDCAN clock source on the H7 Nucleos to PLL2 as suggested but I don't have permission to push the changes to your branch. You can find them [here](https://github.com/chris-durand/modm/tree/h7-fdcan).

Next step for me is to run the unit test on SAM. All other code looks good to merge from my side.

WasabiFan commented:
Thanks! Rebased and pushed. Good catches on the word/byte confusion.

chris-durand commented:
The SAM example works.

chris-durand reviewed (approved):
Thank you very much for doing the message RAM refactoring.

chris-durand commented:
I'm planning to merge this tomorrow.

rleh reviewed (approved):
Really nice work! Thank you a lot 👍🏽
